### PR TITLE
feat(vault): Added an option to use {vault://} in specific fields of plugins

### LINKED ISSUE #14772: kong dbless basic_auth credentials not working with envionment variable
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Kong version (`$ kong version`)

3.7.1

### Current Behavior

Hi Team,

Kong version: 3.7.1 (OSS)
Deployed on AWS EKS version 1.32

I trying to integrate basic_auth credentials as envrionment variable rather hardcoded value in my kong config.
If I use hardcoded value it works and I get 200 OK.
Same value created as K8s secret and integrated as env variable on kong pod throws 401 error.

Here is my config looks like:
dblessConfig:
  vaults:
    - name: env # This name must match the 'env' in vault://env/
      description: "Environment Variable Vault Backend"
  config: |
    _format_version: "1.1"
    consumers:
    - acls:
      - group: monitoring
      basicauth_credentials:
      - password: "{vault://env/MONITORING_BASICAUTH}" 
        username: monitoring
      username: monitoring

The env name: MONITORING_BASICAUTH is present in pod and value is all correct.

My final requirement is to fetch the secret from AWS secret manager and use it as env variable, which partialy working as env variable coming from AWS to k8s pod but using the same didnt work and getting 401 error. 
So I tried creating k8s secret and checked but same result.  

### Expected Behavior

Secret created as env variable should work and give 200 OK. 

### Steps To Reproduce

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by lordgreg**
### Summary

As found in #14772 , we found out, that several plugins, which are using secrets, password or any other secure string, do not support the `{vault://env/*}` templating. This PR adds the support for this feature to the next integrated plugins:
- jwt
- basic-auth
- hmac-auth
- oauth2
- response-transformer
- request-transformer

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/developer.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix #14772

### Additional info for Kong development team

The unit tests are almost the same for the plugins mentioned above with the difference of which fields we actually test. 

I encountered difficulties running the tests in a Dockerized setup (using a custom Dockerfile with all required tools and a docker-compose configuration including Postgres, as well as mounted caches for Cargo, Lua, etc., to avoid reinstalling dependencies on each run). The main issue was related to LuaJIT and running on an ARM-based Mac, which caused frequent test failures such as lj_mem_realloc: allocated memory address 0xffff9588e010 outside required range. This seems to be a LuaJIT-specific problem.

To improve the developer experience, I’d suggest adding a Dockerfile, docker-compose.yaml, and an updated DEVELOPER.md with instructions for running everything inside a Dockerized environment. Please let me know if this would be considered a useful addition — I’d be happy to open a follow-up PR for it.

Please let me know if additional changes are required to the PR.

### CODE REVIEW

**Comment by mschonmeier** [comment]
Will it be possible to access certificates? @lordgreg

**Comment by lordgreg** [comment]
Hi @mschonmeier ,

can you give me an example of configuration or what your use case with certificates is.

**Comment by lordgreg** [comment]
I've fix the linting error being thrown by CICD. 

The other 2 errors arent being thrown because of my code changes.

File: kong/plugins/request-transformer/schema.lua
Comment (raoxiaoyan) [inline kong/plugins/request-transformer/schema.lua]: Why do these fields need to be set as a referenceable field?

**Review by raoxiaoyan — CHANGES_REQUESTED** [review]
For testing all plugins with the vault, please refer [here](https://github.com/Kong/kong/blob/122804286587ed4b238334f635331db2469a93a2/spec/03-plugins/03-http-log/03-schem-vault_spec.lua#L26-L44) and update all of the test cases.

**Comment by lordgreg** [comment]
Thank you for your review @raoxiaoyan . I will complete the requested changes next week.

**Comment by raoxiaoyan** [comment]
@lordgreg Could you add the vault for key-credentials?
https://github.com/Kong/kong/blob/master/kong/plugins/key-auth/daos.lua

**Comment by lordgreg** [comment]
@raoxiaoyan I will add the option too, sure. Which fields you want to be referencable?

**Comment by raoxiaoyan** [comment]
> @raoxiaoyan I will add the option too, sure. Which fields you want to be referencable?

I have asked our PM for this feature. They don't agree to add this option.
Please skip this one. Thank you.

File: kong/plugins/oauth2/daos.lua [hunk 1]
```diff
@@ -31,9 +31,9 @@ local oauth2_credentials = {
     { created_at = typedefs.auto_timestamp_s },
     { consumer = { type = "foreign", reference = "consumers", required = true, on_delete = "cascade", }, },
     { name = { type = "string", required = true }, },
-    { client_id = { type = "string", required = false, unique = true, auto = true }, },
-    { client_secret = { type = "string", required = false, auto = true, encrypted = true }, }, -- encrypted = true is a Kong Enterprise Exclusive feature. It does nothing in Kong CE
-    { hash_secret = { type = "boolean", required = true, default = false }, },
+    { client_id = { type = "string", required = false, unique = true, auto = true, referenceable = true }, },
+    { client_secret = { type = "string", required = false, auto = true, encrypted = true, referenceable = true }, }, -- encrypted = true is a Kong Enterprise Exclusive feature. It does nothing in Kong CE
+    { hash_secret = { type = "boolean", required = true, default = false, referenceable = true }, },
```
[the remark this answers is not in the document]
Comment (lordgreg) [reply kong/plugins/oauth2/daos.lua]: Removed the referenceable from hash_secret.

File: spec/03-plugins/10-basic-auth/06-vault_spec.lua [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  describe("edge cases and validation", function()
+    it("should handle malformed vault references gracefully", function()
+      -- Test various malformed vault references
+      local malformed_refs = {
+        "{vault://invalid/format",
+        "vault://env/missing_braces}",
+        "{vault://env/}",
+        "{vault://env}",
+        "{vault://}",
+        "{vault://env/valid_name?invalid_query=",
```
[the remark this answers is not in the document]
Comment (lordgreg) [reply spec/03-plugins/10-basic-auth/06-vault_spec.lua]: Update the code.

**Comment by lordgreg** [comment]
Hi @raoxiaoyan, @mschonmeier 

the code was updated, the comments were added from my side where ChatGPT gave its opinion. Please let me know if there's anything you require my update on.

**Comment by raoxiaoyan** [comment]
Hi @lordgreg. Thank you for your effort, and appreciate you.
Honestly, the following diagram's test cases do not meet our requirements.
image
Here is an example. Please refers.
https://github.com/Kong/kong/blob/122804286587ed4b238334f635331db2469a93a2/spec/03-plugins/03-http-log/03-schem-vault_spec.lua#L26-L44

**Comment by lordgreg** [comment]
> Hi @lordgreg. Thank you for your effort, and appreciate you. Honestly, the following diagram's test cases do not meet our requirements. image Here is an example. Please refers.
> 
> https://github.com/Kong/kong/blob/122804286587ed4b238334f635331db2469a93a2/spec/03-plugins/03-http-log/03-schem-vault_spec.lua#L26-L44

Thank you for your comment @raoxiaoyan . I've updated all 4 unit tests that are now compliant/same as [kong/spec/03-plugins/03-http-log/03-schem-vault_spec.lua](https://github.com/Kong/kong/blob/122804286587ed4b238334f635331db2469a93a2/spec/03-plugins/03-http-log/03-schem-vault_spec.lua#L26-L44).

File: spec/03-plugins/25-oauth2/06-vault_spec.lua [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    finally(function()
+      helpers.unsetenv(env_name)
+    end)
+
+    helpers.setenv(env_name, env_value)
+
+    local entity = plugins_schema:process_auto_fields({
+      name = PLUGIN_NAME,
+      config = {
+        hide_credentials = "{vault://env/oauth2-hide-credentials}"
```
Comment (raoxiaoyan) [inline spec/03-plugins/25-oauth2/06-vault_spec.lua:39]: You have marked these fields `client_id` and `client_secret` as referenceable.
Right?

File: spec/03-plugins/25-oauth2/06-vault_spec.lua [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    _G.kong = kong_global.new()
+    kong_global.init_pdk(kong, conf)
+
+    local plugin_schema = require("kong.plugins."..PLUGIN_NAME..".schema")
+    assert(plugins_schema:new_subschema(PLUGIN_NAME, plugin_schema))
+  end)
+
+  it("should dereference vault value", function()
+    local env_name = "OAUTH2_HIDE_CREDENTIALS"
+    local env_value = "true"
```
Comment (raoxiaoyan) [inline spec/03-plugins/25-oauth2/06-vault_spec.lua:28]: Please use a proper value for testing.

File: spec/03-plugins/10-basic-auth/06-vault_spec.lua [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    finally(function()
+      helpers.unsetenv(env_name)
+    end)
+
+    helpers.setenv(env_name, env_value)
+
+    local entity = plugins_schema:process_auto_fields({
+      name = PLUGIN_NAME,
+      config = {
+        hide_credentials = "{vault://env/basic-auth-hide-credentials}"
```
Comment (raoxiaoyan) [inline spec/03-plugins/10-basic-auth/06-vault_spec.lua:39]: And also, you have marked `username` and `password` as referenceable.
Please check it again.

**Comment by raoxiaoyan** [comment]
@lordgreg Thank you for updating quickly. I have commented on your newly tested cases. Please update it again for all test cases.

**Comment by lordgreg** [comment]
> http_endpoint

Thank you for your reply @raoxiaoyan. You are right - my problem currently is, that all referenceable fields are written in `daos.lua`, not schema for those plugins, thus the tests need to be updated. I have to add the entries in the db first, then check if values are equal. 

```lua
it("handles vault reference for password with correct hashing", function()
  local env_password = "my_secret_password"
  helpers.setenv("BASIC_AUTH_PASSWORD", env_password)
  
  finally(function()
    helpers.unsetenv("BASIC_AUTH_PASSWORD")
  end)
  
  local consumer = bp.consumers:insert({ username = "testuser" })
  
  local credential, err = db.basicauth_credentials:insert({
    consumer = { id = consumer.id },
    username = "vaultuser",
    password = "{vault://env/basic-auth-password}",
  })
  
  assert.is_nil(err)
  
  -- Verify $refs contains the vault reference
  assert.equal("{vault://env/basic-auth-password}", credential["$refs"].password)
  
  -- Use the same transformation to hash our env value
  local crypto = require "kong.plugins.basic-auth.crypto"
  local expected_hash = crypto.hash(consumer.id, env_password)
  
  -- Verify it matches what's in the database
  assert.equal(expected_hash, credential.password)
end)
```

the same goes for oauth2_credentials, jwt_secrets and hmacauth_credentials (not encrypted). is this the correct approach? please advise.
